AXE benchmarking
================

This is how i'm benchmarking [AXE](https://github.com/kdmurray91/axe). These
benchmarks are designed to show its increased speed and accuracy in certain
situations.

Inputs:
-------

I'm mostly using faked reads ATM.

Eventually!:

I'll use two real world datasets, One set of "pseudoreads" from PARE, which
consist of rescued index reads created by [this script](link needed). The
md5sum of the input file is `<+md5 goes here+>`. This data is unpublished, but
will be put into SRA or similar once the biology paper associated with it is
published.

A second data source is genotyping-by-sequencing reads. These are paired end
reads, however are barcoded in R1 only.
